src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp

```diff
@@ -0,0 +1,390 @@
+    /** adds meta values
+     precursor_in_ms2
+    */
+    defaults_.setValue("topNmatch_fragmenterrors", 7, "If set n > 0,  meta values for topN_meanfragmenterror, topN_MSEfragmenterror, topN_stddevfragmenterror are added");
+    /** adds meta values
+     topN_meanfragmenterror
+     topN_MSEfragmenterror
+     topN_stddevfragmenterror
+    */
+    // TODO right now topN matched, but additional information would be gained if the topN would be assembled from NOT ONLY the matched (high intense peak not identified?)
+
+    defaults_.setValue("fragmenterror_statistics", "true", "If set, meta values for median_fragment_error, IQR_fragment_error are added");
+    defaults_.setValidStrings("fragmenterror_statistics", ListUtils::create<String>("true,false"));
+    /** adds meta values
+     median_fragment_error
+     IQR_fragment_error
+    */
+    defaults_.setValue("terminal_series_match_ratio", "true", "If set, meta values for NTermIonCurrentRatio, CTermIonCurrentRatio are added");
+    defaults_.setValidStrings("terminal_series_match_ratio", ListUtils::create<String>("true,false"));
+    /** adds meta values
+     NTermIonCurrentRatio
+     CTermIonCurrentRatio
+    */
+
+    defaultsToParam_();
+  }
+
+  SpectrumAnnotator::SpectrumAnnotator(const SpectrumAnnotator & rhs) :
+    DefaultParamHandler(rhs)
+  {
+  }
+
+  SpectrumAnnotator & SpectrumAnnotator::operator=(const SpectrumAnnotator & rhs)
+  {
+    if (this != &rhs)
+    {
+      DefaultParamHandler::operator=(rhs);
+    }
+    return *this;
+  }
+
+  SpectrumAnnotator::~SpectrumAnnotator()
+  {
+  }
+
+  MSSpectrum<RichPeak1D> SpectrumAnnotator::annotateMatches(const PeptideHit& ph, const MSSpectrum<Peak1D>& spec, const TheoreticalSpectrumGenerator& tg, const SpectrumAlignment& sa) const
+  {
+    MSSpectrum<RichPeak1D> rich_theoretical_spec;
+    MSSpectrum<RichPeak1D> rich_input_spectrum = MSSpectrumHelper::clone(spec);
+
+    vector<pair<Size, Size> > al;
+    tg.getSpectrum(rich_theoretical_spec, ph.getSequence(), ph.getCharge()-1); //will get y5++, b2+, ... for precursor+++
+    if (!rich_theoretical_spec.isSorted())
+    {
+      rich_theoretical_spec.sortByPosition();
+    }
+    if (!rich_input_spectrum.isSorted())
+    {
+      rich_input_spectrum.sortByPosition();
+    }
+    sa.getSpectrumAlignment(al, rich_theoretical_spec, rich_input_spectrum); //peaks from theor. may be matched to none or one in spec!
+    for (vector<pair<Size, Size > >::const_iterator it = al.begin(); it != al.end(); ++it)
+    {
+        rich_input_spectrum[it->second].setMetaValue("IonMatchError", std::fabs(rich_input_spectrum[it->second].getMZ() - rich_theoretical_spec[it->first].getMZ()));
+        rich_input_spectrum[it->second].setMetaValue("IonName", rich_theoretical_spec[it->first].getMetaValue("IonName"));
+    }
+    Param sap = sa.getParameters();
+    rich_input_spectrum.setMetaValue("fragment_match_tolerance", sap.getValue("tolerance"));
+    return rich_input_spectrum;
```

Comment:
you could also set the unit.
By the way, we previously used: fragment_mass_tolerance and fragment_mass_tolerance_ppm so maybe we can reuse these meta value names